File: backtracking/433-subsets.py

'''
As the Python uses Object-References when passing parameters,
the list (mutable) is passed by References, 
we need to make a deep copy (using copy.deepcopy(cur) or cur[:]) to copy the current subset e.g. cur to answer (list of all subsets).
'''

class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n=len(nums)
        ans=[]
        def dfs(cur,i):
            if i==n:
                ans.append(cur[:])
                return
            
            dfs(cur,i+1)
            dfs(cur+[nums[i]],i+1)
            
        dfs([],0)
        return ans
    def subsets2(self, nums):
        n = len(nums)
        ans = []        
        for i in range(1 << n):
            cur = []
            for j in range(n):
                if i & (1 << j):
                    cur.append(nums[j])
            ans.append(cur)
        return ans   


s=Solution()
print(s.subsets2([1,2,3]))

File: backtracking/433-subsets.cpp
#include<iostream>
#include<vector>

using VEC=std::vector<int>;

using MATRIX=std::vector<std::vector<int>>;
void f(const VEC& nums, MATRIX&  res,VEC& cur,int i)
{
    int n=nums.size();
    if(i==n)
    {
        res.push_back(cur);
        return;
    }
    VEC tmp{cur.begin(),cur.end()};
    f(nums,res,tmp,i+1);
    cur.push_back(nums[i]);
    VEC tmp2{cur.begin(),cur.end()};
    f(nums,res,tmp2,i+1);
}

MATRIX subSets(const VEC& nums)
{
    MATRIX ans{};
    int n=nums.size();
    if(n==0)
        return ans;
    VEC cur;
    f(nums,ans,cur,0);
    return ans;
}

int main()
{
    VEC nums{1,2,3};
    MATRIX res=subSets(nums);
    for(auto &v:res)
    {
        std::cout<<'\n';
        for(auto &d:v)
            std::cout<<d<<',';
    }
}